Calculate the following metrics for each confustion matrix


Confusion Matrix A

| | + | - |
|---|---|---|
|+| 263 | 25 |
|-| 8 | 271|

Accuracy: 0.9417989417989417

Sensitivity: 0.9704797047970479

Specificity: 0.9155405405405406

Precision: 0.9131944444444444

F1: 0.9409660107334525

MCC: 0.8852701097828677

In [1]:
import numpy as np

def confusion_matrix_metrics(true_positives: int, false_positives: int, false_negatives: int, true_negatives: int):
  """Calculates metrics from a confusion matrix.

  Args:
    The four parameters associated with a confusion matrix

  Returns:
    Common metrics for evaluating a model
  """

  accuracy = (true_positives + true_negatives) / (true_positives + true_negatives + false_positives + false_negatives)
  sensitivity = (true_positives) / (true_positives + false_negatives)
  specificity = (true_negatives) / (true_negatives + false_positives)
  f1 = (2 * true_positives) / (2 * true_positives + false_positives + false_negatives)
  mcc = ((true_positives * true_negatives) - (false_positives * false_negatives)) / np.sqrt((true_positives + false_positives)*(true_positives + false_negatives)*(true_negatives + false_positives) *(true_negatives + false_negatives))
  precision = true_positives / (true_positives + false_positives)
  recall = true_positives / (true_positives + false_negatives)

  return accuracy, sensitivity, specificity, precision, f1, mcc

# Calculate precision and recall.
accuracy, sensitivity, specificity, precision, f1, mcc = confusion_matrix_metrics(263, 25, 8, 271)

# Print precision and recall.
print("Accuracy:", accuracy)
print("Sensitivity:", sensitivity)
print("Specificity:", specificity)
print("Precision:",precision)
print("F1:", f1)
print("MCC:", mcc)

Accuracy: 0.9417989417989417
Sensitivity: 0.9704797047970479
Specificity: 0.9155405405405406
Precision: 0.9131944444444444
F1: 0.9409660107334525
MCC: 0.8852701097828677


Confusion Matrix B

| | + | - |
|---|---|---|
|+| 176 | 112 |
|-| 33 | 246|

Accuracy: 0.7442680776014109

Sensitivity: 0.8421052631578947

Specificity: 0.6871508379888268

Precision: 0.6111111111111112

F1: 0.7082494969818913

MCC: 0.5107191988044958




In [2]:
# Calculate precision and recall.
accuracy, sensitivity, specificity, precision, f1, mcc = confusion_matrix_metrics(176, 112, 33, 246)

# Print precision and recall.
print("Accuracy:", accuracy)
print("Sensitivity:", sensitivity)
print("Specificity:", specificity)
print("Precision:",precision)
print("F1:", f1)
print("MCC:", mcc)

Accuracy: 0.7442680776014109
Sensitivity: 0.8421052631578947
Specificity: 0.6871508379888268
Precision: 0.6111111111111112
F1: 0.7082494969818913
MCC: 0.5107191988044958


Confusion Matrix C

| | + | - |
|---|---|---|
|+| 18 | 42 |
|-| 3 | 271|

Accuracy: 0.8652694610778443

Sensitivity: 0.8571428571428571

Specificity: 0.865814696485623

Precision: 0.3

F1: 0.4444444444444444

MCC: 0.45713419521588783


In [3]:
# Calculate precision and recall.
accuracy, sensitivity, specificity, precision, f1, mcc = confusion_matrix_metrics(18, 42, 3, 271)

# Print precision and recall.
print("Accuracy:", accuracy)
print("Sensitivity:", sensitivity)
print("Specificity:", specificity)
print("Precision:",precision)
print("F1:", f1)
print("MCC:", mcc)

Accuracy: 0.8652694610778443
Sensitivity: 0.8571428571428571
Specificity: 0.865814696485623
Precision: 0.3
F1: 0.4444444444444444
MCC: 0.45713419521588783


Confusion Matrix D

| | + | - |
|---|---|---|
|+| 35 | 25 |
|-| 20 | 254|

Accuracy: 0.8652694610778443

Sensitivity: 0.6363636363636364

Specificity: 0.910394265232975

Precision: 0.5833333333333334

F1: 0.6086956521739131

MCC: 0.5282355283220976


In [4]:
# Calculate precision and recall.
accuracy, sensitivity, specificity, precision, f1, mcc = confusion_matrix_metrics(35, 25, 20, 254)

# Print precision and recall.
print("Accuracy:", accuracy)
print("Sensitivity:", sensitivity)
print("Specificity:", specificity)
print("Precision:",precision)
print("F1:", f1)
print("MCC:", mcc)

Accuracy: 0.8652694610778443
Sensitivity: 0.6363636363636364
Specificity: 0.910394265232975
Precision: 0.5833333333333334
F1: 0.6086956521739131
MCC: 0.5282355283220976


Use the sklearn algorithms and builtin datasets to answer the following questions

How many instances are in the diabetes dataset?  Use code to show your answer
442

In [ ]:
import pandas as pd
from sklearn.datasets import load_diabetes

diabetes_data = load_diabetes()

df = pd.DataFrame(data=diabetes_data.data, columns=diabetes_data.feature_names)
# df.info()
print("Number of instances:", len(df))

Number of instances: 442


How many instances are in the iris dataset? Use code to show your answer
150

In [ ]:
import pandas as pd
from sklearn.datasets import load_iris

iris_data = load_iris()

df = pd.DataFrame(data=iris_data.data, columns=iris_data.feature_names)
# df.info()
print("Number of instances:", len(df))

Number of instances: 150


Use the iris dataset and the digits dataset in conjuction with the DecisionTreeClassifier, the GuassianNB, and the KNeighborsClassifier to train and test each classifier on each dataset.  Report the accuracy, precision, recall, and F1 scores.


In [27]:
from sklearn.datasets import load_iris, load_digits
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
import pandas as pd

def process_models(x, y):
  # Split the data into training and test
  x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=13)

  # Train DescisionTreeClassifier
  model = DecisionTreeClassifier()
  model.fit(x_train, y_train)
  predictions = model.predict(x_test)
  print("DescisionTreeClassifier")
  print_metrics(y_test, predictions)

  # Train GaussianNB
  model = GaussianNB()
  model.fit(x_train, y_train)
  predictions = model.predict(x_test)
  print("GaussianNB")
  print_metrics(y_test, predictions)

  # Train KNeighborsClassifier
  model = KNeighborsClassifier()
  model.fit(x_train, y_train)
  predictions = model.predict(x_test)
  print("KNeighborsClassifier")
  print_metrics(y_test, predictions)

def print_metrics(y_test, predictions):
  print("Accuracy:", accuracy_score(y_test, predictions))
  print("Sensitivity/Recall:", recall_score(y_test, predictions, average='weighted'))
  print("Precision:", precision_score(y_test, predictions, average='weighted'))
  print("F1:", f1_score(y_test, predictions, average='weighted'))
  print("\n")


# Load Data
toydata = load_iris()
x, y = toydata.data, toydata.target
print("**IRIS MODELS**")
process_models(x, y)

# Load Data
toydata = load_digits()
x, y = toydata.data, toydata.target
print("**DIGITS MODELS**")
process_models(x, y)

**IRIS MODELS**
DescisionTreeClassifier
Accuracy: 0.9333333333333333
Sensitivity/Recall: 0.9333333333333333
Precision: 0.9466666666666667
F1: 0.9342592592592592


GaussianNB
Accuracy: 0.9666666666666667
Sensitivity/Recall: 0.9666666666666667
Precision: 0.9703703703703703
F1: 0.9669803921568628


KNeighborsClassifier
Accuracy: 0.9
Sensitivity/Recall: 0.9
Precision: 0.9272727272727274
F1: 0.9013729977116705


**DIGITS MODELS**
DescisionTreeClassifier
Accuracy: 0.8833333333333333
Sensitivity/Recall: 0.8833333333333333
Precision: 0.8873229491128701
F1: 0.8841906730445793


GaussianNB
Accuracy: 0.8472222222222222
Sensitivity/Recall: 0.8472222222222222
Precision: 0.8698056241176428
F1: 0.8500110465276695


KNeighborsClassifier
Accuracy: 0.9833333333333333
Sensitivity/Recall: 0.9833333333333333
Precision: 0.9842757936507937
F1: 0.9833504312814445




**IRIS MODELS**

DescisionTreeClassifier

Accuracy: 0.9333333333333333

Sensitivity/Recall: 0.9333333333333333

Precision: 0.9466666666666667

F1: 0.9342592592592592




GaussianNB

Accuracy: 0.9666666666666667

Sensitivity/Recall: 0.9666666666666667

Precision: 0.9703703703703703

F1: 0.9669803921568628



KNeighborsClassifier

Accuracy: 0.9

Sensitivity/Recall: 0.9

Precision: 0.9272727272727274

F1: 0.9013729977116705




**DIGITS MODELS**

DescisionTreeClassifier

Accuracy: 0.8833333333333333

Sensitivity/Recall: 0.8833333333333333

Precision: 0.8873229491128701

F1: 0.8841906730445793




GaussianNB

Accuracy: 0.8472222222222222

Sensitivity/Recall: 0.8472222222222222

Precision: 0.8698056241176428

F1: 0.8500110465276695




KNeighborsClassifier

Accuracy: 0.9833333333333333

Sensitivity/Recall: 0.9833333333333333

Precision: 0.9842757936507937

F1: 0.9833504312814445


